// @ionic-super-tabs/core/dist/collection/super-tab-button/super-tab-button.component.js
export class SuperTabButtonComponent {
    constructor() {
        this.scrollableContainer = false;
    }
    indexChildren() {
        this.label = this.el.querySelector('ion-label');
        this.icon = this.el.querySelector('ion-icon');
    }
    componentWillLoad() {
        this.indexChildren();
    }
    componentDidUpdate() {
        this.indexChildren();
    }
    hostData() {
        return {
            role: 'button',
            'aria-label': this.label ? this.label.innerText : false,
            'aria-disabled': this.disabled ? 'true' : false,
            class: {
                'ion-activatable': true,
                'ion-focusable': true,
                'icon-only': !!this.icon && !this.label,
                'label-only': !!this.label && !this.icon,
                active: Boolean(this.active),
                scrollableContainer: this.scrollableContainer,
            },
        };
    }
    render() {
        return [
            h("slot", null),
            h("ion-ripple-effect", { type: "unbounded" }),
        ];
    }
    static get is() { return "super-tab-button"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "active": {
            "type": Boolean,
            "attr": "active",
            "reflectToAttr": true
        },
        "disabled": {
            "type": Boolean,
            "attr": "disabled",
            "reflectToAttr": true
        },
        "el": {
            "elementRef": true
        },
        "icon": {
            "state": true
        },
        "index": {
            "type": Number,
            "attr": "index",
            "reflectToAttr": true
        },
        "label": {
            "state": true
        },
        "scrollableContainer": {
            "type": Boolean,
            "attr": "scrollable-container"
        }
    }; }
    static get style() { return "/**style-placeholder:super-tab-button:**/"; }
}
