// @ionic-super-tabs/core/dist/cjs/es5/build/chunk-39d8d03b.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DEFAULT_CONFIG = { dragThreshold: 10, allowElementScroll: !1, maxDragAngle: 40, sideMenuThreshold: 50, transitionDuration: 300, shortSwipeDuration: 300 };
exports.a = DEFAULT_CONFIG;
function pointerCoord(r) { if (r) {
    var o = r.changedTouches;
    if (o && o.length > 0) {
        var t = o[0];
        return { x: t.clientX, y: t.clientY };
    }
    if (void 0 !== r.pageX)
        return { x: r.pageX, y: r.pageY };
} return { x: 0, y: 0 }; }
exports.c = pointerCoord;
var easeInOutCubic = function (r) { return r < .5 ? 4 * r * r * r : (r - 1) * (2 * r - 2) * (2 * r - 2) + 1; };
function getScrollCoord(r, o, t, e, n) { var l = Math.min(1, (e - t) / n), a = easeInOutCubic(l); return Math.ceil(a * (o - r) + r); }
function scroll(r, o, t, e, n, l, a, c) { var i = window.performance.now(), s = o === e ? e : getScrollCoord(o, e, l, i, a), u = t === n ? n : getScrollCoord(t, n, l, i, a); r.scrollTo(s, u), i - l >= a || c.write(function () { scroll(r, o, t, e, n, l, a, c); }); }
var scrollEl = function (r, o, t, e, n) { void 0 === e && (e = 300), e <= 0 ? n.write(function () { r.scrollTo(o, t); }) : n.read(function () { var l = r.scrollLeft, a = r.scrollTop, c = window.performance.now(); n.write(function () { scroll(r, l, a, o, t, c, e, n); }); }); };
exports.b = scrollEl;
function checkGesture(r, o, t) { if (o) {
    var e = t.maxDragAngle * (Math.PI / 180), n = Math.cos(e), l = r.x - o.x, a = r.y - o.y;
    if (Math.sqrt(l * l + a * a) >= t.dragThreshold) {
        var c = Math.atan2(a, l), i = Math.cos(c);
        return Math.abs(i) > n;
    }
} }
exports.d = checkGesture;
function getScrollX(r, o) { return r.scrollLeft + ("number" == typeof o ? o : 0); }
function getNormalizedScrollX(r, o) { var t = r.scrollWidth - r.clientWidth, e = getScrollX(r, o); return Math.max(0, Math.min(t, e)); }
exports.e = getNormalizedScrollX;
