// @ionic-super-tabs/core/dist/esm/es2017/build/vexfz1ul.entry.js
import { h } from '../supertabs.core.js';

import { b as scrollEl, c as pointerCoord, d as checkGesture, e as getNormalizedScrollX } from './chunk-39d8d03b.js';

class SuperTabButtonComponent {
    constructor() {
        this.scrollableContainer = false;
    }
    indexChildren() {
        this.label = this.el.querySelector('ion-label');
        this.icon = this.el.querySelector('ion-icon');
    }
    componentWillLoad() {
        this.indexChildren();
    }
    componentDidUpdate() {
        this.indexChildren();
    }
    hostData() {
        return {
            role: 'button',
            'aria-label': this.label ? this.label.innerText : false,
            'aria-disabled': this.disabled ? 'true' : false,
            class: {
                'ion-activatable': true,
                'ion-focusable': true,
                'icon-only': !!this.icon && !this.label,
                'label-only': !!this.label && !this.icon,
                active: Boolean(this.active),
                scrollableContainer: this.scrollableContainer,
            },
        };
    }
    render() {
        return [
            h("slot", null),
            h("ion-ripple-effect", { type: "unbounded" }),
        ];
    }
    static get is() { return "super-tab-button"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "active": {
            "type": Boolean,
            "attr": "active",
            "reflectToAttr": true
        },
        "disabled": {
            "type": Boolean,
            "attr": "disabled",
            "reflectToAttr": true
        },
        "el": {
            "elementRef": true
        },
        "icon": {
            "state": true
        },
        "index": {
            "type": Number,
            "attr": "index",
            "reflectToAttr": true
        },
        "label": {
            "state": true
        },
        "scrollableContainer": {
            "type": Boolean,
            "attr": "scrollable-container"
        }
    }; }
    static get style() { return ":host{-ms-flex:1 0 0px;flex:1 0 0;cursor:pointer;position:relative;max-width:100%;overflow:hidden;display:-ms-flexbox;display:flex;-ms-flex-flow:column nowrap;flex-flow:column nowrap;-ms-flex-align:center;align-items:center;-ms-flex-pack:justify;justify-content:space-between;-webkit-touch-callout:none;-webkit-box-sizing:border-box;box-sizing:border-box;-webkit-transform:translateZ(0);transform:translateZ(0);opacity:.7;height:72px;padding:12px 16px 20px 16px;--st-icon-size:24px;--st-label-line-height:14px;--st-label-height:14px;--st-label-font-size:14px;--st-label-text-transform:uppercase;--st-label-font-weight:500}::slotted{-ms-flex-negative:1;flex-shrink:1}::slotted(ion-label){color:var(--st-label-color,var(--ion-color-contrast));line-height:var(--st-label-line-height);height:var(--st-label-height);font-size:var(--st-label-font-size);text-transform:var(--st-label-text-transform);font-weight:var(--st-label-font-weight)}::slotted(ion-icon){color:var(--st-icon-color,var(--ion-color-contrast));min-height:var(--st-icon-size);min-width:var(--st-icon-size);font-size:var(--st-icon-size)}:host(.active){opacity:1}:host(.icon-only),:host(.label-only){padding:12px 16px;height:48px;-ms-flex-pack:center;justify-content:center}:host(.scrollableContainer){-ms-flex-positive:0;flex-grow:0;-ms-flex-preferred-size:auto;flex-basis:auto;min-width:90px;max-width:360px}"; }
}

class SuperTabsToolbarComponent {
    constructor() {
        this.showIndicator = true;
        this.color = 'primary';
        this.scrollable = false;
        this.scrollablePadding = true;
        this.activeTabIndex = 0;
        this.leftThreshold = 0;
        this.rightThreshold = 0;
    }
    componentDidLoad() {
        this.indexButtons();
    }
    componentDidUpdate() {
        this.indexButtons();
    }
    setActiveTab(index) {
        this.activeTabIndex = index;
        this.alignIndicator(index, true);
        this.markButtonActive(this.buttons[index]);
    }
    setSelectedTab(index) {
        this.alignIndicator(index);
    }
    moveContainer(scrollX, animate) {
        scrollEl(this.buttonsContainerEl, scrollX, 0, animate ? this.config.transitionDuration : 0, this.queue);
    }
    onWindowResize() {
        this.alignIndicator(this.activeTabIndex);
    }
    onClick(ev) {
        let button = ev.target;
        const tag = button.tagName.toLowerCase();
        if (tag !== 'super-tab-button') {
            if (tag === 'super-tabs-toolbar') {
                return;
            }
            button = button.closest('super-tab-button');
        }
        this.setActiveTab(button.index);
        this.buttonClick.emit(button);
    }
    async onTouchStart(ev) {
        this.queue.read(() => {
            const coords = pointerCoord(ev);
            const vw = this.el.clientWidth;
            if (coords.x < this.leftThreshold || coords.x > vw - this.rightThreshold) {
                this.shouldCapture = false;
                return;
            }
            this.initialCoords = coords;
            this.lastPosX = coords.x;
        });
    }
    async onTouchMove(ev) {
        this.queue.read(() => {
            const coords = pointerCoord(ev);
            if (typeof this.lastPosX !== 'number') {
                return;
            }
            if (!this.isDragging) {
                if (typeof this.shouldCapture !== 'boolean') {
                    this.shouldCapture = checkGesture(coords, this.initialCoords, this.config);
                }
                if (this.shouldCapture !== true) {
                    return;
                }
                this.isDragging = true;
            }
            const deltaX = this.lastPosX - coords.x;
            this.lastPosX = coords.x;
            if (deltaX === 0) {
                return;
            }
            const scrollLeft = this.buttonsContainerEl.scrollLeft;
            const scrollX = getNormalizedScrollX(this.buttonsContainerEl, deltaX);
            if (scrollX === scrollLeft) {
                return;
            }
            this.moveContainer(scrollX, false);
        });
    }
    async onTouchEnd() {
        this.isDragging = false;
        this.shouldCapture = void 0;
        this.lastPosX = void 0;
    }
    indexButtons() {
        const buttons = this.el.querySelectorAll('super-tab-button');
        const buttonsArray = [];
        for (let i = 0; i < buttons.length; i++) {
            const button = buttons[i];
            button.index = i;
            button.scrollableContainer = this.scrollable;
            if (this.activeTabIndex === i) {
                this.markButtonActive(button);
            }
            buttonsArray.push(button);
        }
        this.buttons = buttonsArray;
        this.alignIndicator(this.activeTabIndex);
        if (this.config.sideMenu === 'both' || this.config.sideMenu === 'left') {
            this.leftThreshold = this.config.sideMenuThreshold;
        }
        if (this.config.sideMenu === 'both' || this.config.sideMenu === 'right') {
            this.rightThreshold = this.config.sideMenuThreshold;
        }
    }
    markButtonActive(button) {
        if (this.activeButton) {
            this.activeButton.active = false;
        }
        button.active = true;
        this.activeButton = button;
    }
    adjustContainerScroll(animate) {
        this.queue.read(() => {
            let pos;
            const ip = this.indicatorPosition;
            const iw = this.indicatorWidth;
            const mw = this.buttonsContainerEl.clientWidth;
            const sp = this.buttonsContainerEl.scrollLeft;
            const centerDelta = (mw / 2 - iw / 2);
            if (ip + iw + centerDelta > mw + sp) {
                const delta = ip + iw + centerDelta - mw - sp;
                pos = sp + delta;
            }
            else if (ip - centerDelta < sp) {
                pos = ip - centerDelta;
                pos = Math.max(pos, 0);
                pos = pos > ip ? ip - mw + iw : pos;
            }
            if (typeof pos === 'number') {
                this.moveContainer(pos, animate);
            }
        });
    }
    alignIndicator(index, animate = false) {
        if (!this.showIndicator) {
            return;
        }
        this.queue.read(() => {
            const remainder = index % 1;
            const isDragging = this.isDragging = remainder > 0;
            let position, width;
            const floor = Math.floor(index), ceil = Math.ceil(index);
            const button = this.buttons[floor];
            position = button.offsetLeft;
            width = button.clientWidth;
            if (this.isDragging && floor !== ceil) {
                const buttonB = this.buttons[ceil];
                const buttonBWidth = buttonB.clientWidth;
                const buttonBPosition = buttonB.offsetLeft;
                position += remainder * (buttonBPosition - position);
                width += remainder * (buttonBWidth - width);
            }
            this.indicatorPosition = position;
            this.indicatorWidth = width;
            this.adjustContainerScroll(animate || !isDragging);
            this.queue.write(() => {
                if (!this.showIndicator || this.indicatorEl) {
                    this.indicatorEl.style.setProperty('--st-indicator-position-x', this.indicatorPosition + 'px');
                    this.indicatorEl.style.setProperty('--st-indicator-scale-x', String(this.indicatorWidth / 100));
                    this.indicatorEl.style.setProperty('--st-indicator-transition-duration', this.isDragging ? '0' : `${this.config.transitionDuration}ms`);
                }
            });
        });
    }
    hostData() {
        return {
            class: {
                ['ion-color-' + this.color]: true,
            },
        };
    }
    render() {
        return [
            h("div", { class: "buttons-container", ref: (ref) => this.buttonsContainerEl = ref },
                h("slot", null),
                this.showIndicator &&
                    h("super-tab-indicator", { ref: (ref) => this.indicatorEl = ref, toolbarPosition: this.el.assignedSlot.name })),
        ];
    }
    static get is() { return "super-tabs-toolbar"; }
    static get encapsulation() { return "shadow"; }
    static get properties() { return {
        "color": {
            "type": String,
            "attr": "color"
        },
        "config": {
            "type": "Any",
            "attr": "config",
            "mutable": true
        },
        "el": {
            "elementRef": true
        },
        "moveContainer": {
            "method": true
        },
        "queue": {
            "context": "queue"
        },
        "scrollable": {
            "type": Boolean,
            "attr": "scrollable",
            "reflectToAttr": true
        },
        "scrollablePadding": {
            "type": Boolean,
            "attr": "scrollable-padding",
            "reflectToAttr": true
        },
        "setActiveTab": {
            "method": true
        },
        "setSelectedTab": {
            "method": true
        },
        "showIndicator": {
            "type": Boolean,
            "attr": "show-indicator"
        }
    }; }
    static get events() { return [{
            "name": "buttonClick",
            "method": "buttonClick",
            "bubbles": true,
            "cancelable": true,
            "composed": true
        }]; }
    static get listeners() { return [{
            "name": "window:resize",
            "method": "onWindowResize",
            "passive": true
        }, {
            "name": "click",
            "method": "onClick"
        }, {
            "name": "touchstart",
            "method": "onTouchStart",
            "passive": true
        }, {
            "name": "touchmove",
            "method": "onTouchMove",
            "passive": true
        }, {
            "name": "touchend",
            "method": "onTouchEnd",
            "passive": true
        }]; }
    static get style() { return ":host{-ms-flex:0 1 auto;flex:0 1 auto;display:block;position:relative;z-index:2;background:var(--super-tabs-toolbar-background,var(--ion-color-base));overflow:visible}:host,:host .buttons-container{width:100%;-webkit-transform:translateZ(0);transform:translateZ(0)}:host .buttons-container{display:-ms-flexbox;display:flex;-ms-flex-flow:row nowrap;flex-flow:row nowrap;overflow:hidden}:host(:not([no-border])):after{left:0;bottom:-8px;background-position:left 0 top 0;position:absolute;width:100%;height:8px;background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAIBAMAAAACWGKkAAAAFVBMVEUAAAAAAAAAAAAAAAAAAAAAAAAAAAASAQCkAAAAB3RSTlMFTEIzJBcOYhQUIwAAAB9JREFUCNdjEIQCBiUoYDCGAgYXKGAIhQKGNChgwAAAorMLKSCkL40AAAAASUVORK5CYII=);background-repeat:repeat-x;content:\"\";z-index:3;-webkit-transform:translateZ(0);transform:translateZ(0)}:host([scrollable]) .buttons-container{will-change:scroll-position}:host([scrollable][scrollable-padding]) .buttons-container{padding-left:var(--st-scrollable-toolbar-padding-left,52px);width:calc(100% - var(--st-scrollable-toolbar-padding-left, 52px))}"; }
}

export { SuperTabButtonComponent as SuperTabButton, SuperTabsToolbarComponent as SuperTabsToolbar };
